### PR TITLE
Mixed precision astronomy

### PR DESCRIPTION
**Pull request by mcallic2**
**Description**
Changes to add mixed precision to astronomy files

Fixes # (issue)

**How Has This Been Tested?**
Autotools using oneapi and cmake

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: astronomy/include/astronomy.inc [hunk 1]
```diff
@@ -854,122 +183,142 @@ real, dimension(:,:), intent(out), optional :: half_day_out
...
-      t = gmt + lon - PI
-      where(t >= PI) t = t - twopi
-      where(t < -PI) t = t + twopi
 
-      Lallow_negative = .false.
-      if (present(allow_negative_cosz)) then
-         if (allow_negative_cosz) Lallow_negative = .true.
-      end if
+    t = gmt + lon - PI
+    where(t >= PI) t = real(t, FMS_AST_KIND_) - real(twopi, FMS_AST_KIND_)
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:218]: t is already in FMS_AST_KIND_

File: astronomy/include/astronomy.inc [hunk 2]
```diff
@@ -854,122 +183,142 @@ real, dimension(:,:), intent(out), optional :: half_day_out
...
 !--------------------------------------------------------------------
-      t = gmt + lon - PI
-      where(t >= PI) t = t - twopi
-      where(t < -PI) t = t + twopi
 
-      Lallow_negative = .false.
-      if (present(allow_negative_cosz)) then
-         if (allow_negative_cosz) Lallow_negative = .true.
-      end if
+    t = gmt + lon - PI
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:217]: real(PI,FMS_AST_KIND_)

File: astronomy/include/astronomy.inc [hunk 3]
```diff
@@ -854,122 +183,142 @@ real, dimension(:,:), intent(out), optional :: half_day_out
...
 !    case 8: averaging period begins after sunset and ends after the
 !    next day's sunrise but before the next day's sunset. if the
 !    averaging period is less than a half-day (pi) the latter
 !    circumstance will never occur.
 !-----------------------------------------------------------------
-        where(  h <  t .and. twopi - h < tt  )
-          cosz = aa + bb*(stt + sh) / (tt + h - twopi)
+
+        where (h <  t .and. real(twopi, FMS_AST_KIND_) - h < tt)
+            cosz = real(aa, FMS_AST_KIND_) + real(bb, FMS_AST_KIND_)&
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:315]: why all the real(,FMS_AST_KIND_) here?

File: astronomy/include/astronomy.inc [hunk 4]
```diff
@@ -978,45 +327,49 @@ real, dimension(:,:), intent(out), optional :: half_day_out
...
-                                                         twopi)/dt
+
+    where (t < -h .and. tt < -h)           fracday = 0.0_lkind
+    where (t < -h .and. abs(tt) <= h)      fracday = (tt + h )/dt
+    where (t < -h .and. h < tt)            fracday = ( h + h )/dt
+    where (abs(t) <= h .and. abs(tt) <= h) fracday = (tt - t )/dt
+    where (abs(t) <= h .and. h < tt)       fracday = ( h - t )/dt
+    where (h <  t)                         fracday = 0.0_lkind
+    where (real(twopi, FMS_AST_KIND_) - h < tt)  &
+                                           fracday = real(fracday, FMS_AST_KIND_) + &
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:338]: here too, fracday does not need real(,FMS_AST_KIND_)

File: astronomy/include/astronomy.inc [hunk 5]
```diff
@@ -1378,52 +756,57 @@ end subroutine diurnal_solar_cal_0d
...
-      h_out = h/PI
-      rr_out = rr
 
-end subroutine daily_mean_solar_2d
+    where (h == 0.0_lkind)
+          cosz = 0.0_lkind
+    elsewhere
+          cosz = sin(lat)*sin(dec) + cos(lat)*cos(dec)*sin(h)/h
+    end where
+      h_out    = h/PI
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:806]: real(PI,FMS_AST_KIND_)

File: astronomy/include/astronomy.inc [hunk 6]
```diff
@@ -1740,50 +1153,56 @@ end subroutine daily_mean_solar_cal_0d
...
 !!    orbital increments.
 !----------------------------------------------------------------------
-        solar = 0.0
-        cosz = 0.0
+
+        solar   = 0.0_lkind
+        cosz    = 0.0_lkind
         do n =1, num_angles
-          t = float(n-1)*twopi/float(num_angles)
+          t     = real(float(n-1), FMS_AST_KIND_)*real(twopi, FMS_AST_KIND_)/real(float(num_angles), FMS_AST_KIND_)
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1191]: don't need the float here

File: astronomy/include/astronomy.inc [hunk 7]
```diff
@@ -1740,50 +1153,56 @@ end subroutine daily_mean_solar_cal_0d
...
           call daily_mean_solar(lat,t, z, fracday, rrsun)
-          s = z*rrsun*fracday
+          s     = z*rrsun*fracday
           solar = solar + s
           cosz  = cosz  + z*s
         end do
-        solar = solar/float(num_angles)
-        cosz  = cosz/float(num_angles)
+        solar   = solar/float(num_angles)
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1197]: real(num_angles,FMS_AST_KIND_)

File: astronomy/include/astronomy.inc [hunk 8]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
-      norm_time = t*float(num_angles)/twopi
-      int = floor(norm_time)
-      int = modulo(int,num_angles)
-      int_1 = int+1
-      x = norm_time - floor(norm_time)
-      angle = (1.0 -x)*orb_angle(int) + x*orb_angle(int_1)
-      angle = modulo(angle, twopi)
 
-end function angle
+    norm_time = real(t, FMS_AST_KIND_)*real(float(num_angles), FMS_AST_KIND_)/real(twopi, FMS_AST_KIND_)
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1469]: don't need float here

File: astronomy/include/astronomy.inc [hunk 9]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
+    int       = floor(norm_time)
+    int       = modulo(int,num_angles)
+    int_1     = int+1
+    x         = norm_time - floor(norm_time)
+    ANGLE_    = (1.0_lkind - real(x, FMS_AST_KIND_))*real(orb_angle(int), FMS_AST_KIND_) &
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1474]: x is already in FMS_AST_KIND_

File: astronomy/include/astronomy.inc [hunk 10]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
+                + real(x, FMS_AST_KIND_)*real(orb_angle(int_1), FMS_AST_KIND_)
+    ANGLE_    = modulo(real(ANGLE_, FMS_AST_KIND_), real(twopi, FMS_AST_KIND_))
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1476]: ANGLE_ is already in FMS_AST_KIND_

File: astronomy/include/astronomy.inc [hunk 11]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
 !---------------------------------------------------------------------
-      real, dimension (size(latitude,1),size(latitude,2)):: &
+
+real(kind=FMS_AST_KIND_), dimension (size(latitude,1),size(latitude,2)):: &
                                                   cos_half_day, & !< Cosine of half-day length [dimensionless]
                                                   lat !< Model latitude, adjusted so that it is never 0.5*pi or -0.5*pi
-      real :: tan_dec !< tangent of solar declination [dimensionless]
-      real :: eps = 1.0E-05 !< small increment
+real(kind=FMS_AST_KIND_) :: tan_dec !< tangent of solar declination [dimensionless]
+real(kind=FMS_AST_KIND_) :: eps = real(1.0E-05, FMS_AST_KIND_) !< small increment
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1530]: 1.0E-5_lkind

File: astronomy/include/astronomy.inc [hunk 12]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
 !--------------------------------------------------------------------
 !>    adjust latitude so that its tangent will be defined.
 !--------------------------------------------------------------------
-      lat = latitude
-      where (latitude ==  0.5*PI) lat= latitude - eps
-      where (latitude == -0.5*PI) lat= latitude + eps
+
+    lat = latitude
+    where (latitude == 0.5_lkind*PI)  lat = latitude - eps
+    where (latitude == -0.5_lkind*PI) lat = latitude + eps
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1545]: real(PI,FMS_AST_KIND_)

File: astronomy/include/astronomy.inc [hunk 13]
```diff
@@ -2092,85 +1465,97 @@ real, intent(in) :: t !< time of year (between 0 and 2*pi; t=0 at NH autumnal eq
...
 !--------------------------------------------------------------------
-      cos_half_day = -tan(lat)*tan_dec
-      where (cos_half_day <= -1.0)  h = PI
-      where (cos_half_day >= +1.0)  h = 0.0
-      where(cos_half_day > -1.0 .and. cos_half_day < 1.0) &
-                                               h = acos(cos_half_day)
 
-end function half_day_2d
+    cos_half_day = -tan(lat)*tan_dec
+    where (cos_half_day <= -1.0_lkind) h = PI
```
Comment (mlee03) [inline astronomy/include/astronomy.inc:1553]: real(PI,FMS_AST_KIND_)

**Review by mlee03 — CHANGES_REQUESTED** [review]

File: astronomy/astronomy.F90 [hunk 14]
```diff
@@ -56,152 +51,37 @@ module astronomy_mod
...
-!> @ingroup astronomy_mod
+   module procedure diurnal_solar_2d_r4, diurnal_solar_2d_r8
+   module procedure diurnal_solar_1d_r4, diurnal_solar_1d_r8
+   module procedure diurnal_solar_0d_r4, diurnal_solar_0d_r8
+   module procedure diurnal_solar_cal_2d_r4, diurnal_solar_cal_2d_r8
+   module procedure diurnal_solar_cal_1d_r4, diurnal_solar_cal_1d_r8
+   module procedure diurnal_solar_cal_0d_r4, diurnal_solar_cal_0d_r8
+end interface diurnal_solar
+
+
```
Comment (mlee03) [inline astronomy/astronomy.F90:74]: Should there be more interfaces for all the subroutines and functions in astronomy.inc?

File: astronomy/astronomy.F90 [hunk 15]
```diff
@@ -774,1476 +587,132 @@ subroutine get_ref_date_of_ae (day_out,month_out,year_out,&
...
-!  local variables:
+!>    allocate the orbital angle array, sized by the namelist parameter
+!!    num_angles, defining the annual cycle resolution of the earth's
+!!    orbit. define some constants to be used.
 !--------------------------------------------------------------------
-      real, dimension(1,1) :: lat_2d, lon_2d, cosz_2d, fracday_2d, halfday_2d
+! wfc moving to astronomy_init
+!     allocate ( orb_angle(0:num_angles) )
+orb_angle(0) = 0.0_r8_kind
+dt = twopi/float(num_angles)
```
[the remark this answers is not in the document]
Comment (mcallic2) [reply astronomy/astronomy.F90:657]: resolved in 9e832558

[the remark this answers is not in the document]
Comment (mcallic2) [reply astronomy/astronomy.F90]: resolved in 9e832558abec513634a54b05e2d114b93414ccad

**Comment by mlee03** [comment]
@mcallic2 , could you add @J-Lentz as a reviewer

File: test_fms/astronomy/test_daily_mean.F90 [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+    real(kind=TEST_AST_KIND_), dimension(n,n) :: cosz_2d, h_out_2d
+    real(kind=TEST_AST_KIND_), dimension(n,n) :: ref_cosz_2d, ref_h_out_2d
+    real(kind=TEST_AST_KIND_)                 :: rr_out, ref_rr_out, ref_dec
+    integer                                   :: i, j, counter
+
+    ! h_2d is calculated from the private function HALF_DAY_2D
+    h_2d = real(PI,TEST_AST_KIND_)/2.0_lkind
+
+    counter = 1
+    do i = 1, 4
```
Comment (mlee03) [inline test_fms/astronomy/test_daily_mean.F90:81]: could use n instead of 4

File: test_fms/astronomy/test_annual_mean_solar_1d.F90 [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    do i = 1, 16
+      lat(i) = real(i, TEST_AST_KIND_) * real(PI,TEST_AST_KIND_)/8.0_lkind
+    end do
+
+    call create_ref_values_1d(jst, jnd, lat, ref_cosz, ref_solar, ref_fracday, ref_rrsun)
+    call annual_mean_solar(jst, jnd, lat, cosz, solar, fracday, rrsun)
+
+    do i = 1, 16
+
+      if (abs(ref_cosz(i) - cosz(i)) .gt. real(1E-07,TEST_AST_KIND_)) then
```
Comment (mlee03) [inline test_fms/astronomy/test_annual_mean_solar_1d.F90:117]: 1.e-7_lkind

**Review by mlee03 — COMMENTED** [review]
@mcallic2 make sure to compile with  -Wall -Wconversion-extra for GCC to catch all the kind/type mismatch

File: astronomy/astronomy.F90 [hunk 18]
```diff
@@ -519,20 +561,19 @@ subroutine get_period_integer (period_out)
...
+    if (.not. module_is_initialized)   &
+        call error_mesg ( 'astronomy_mod', ' module has not been initialized', FATAL)
 
 !--------------------------------------------------------------------
 !    define length of year in seconds.
 !--------------------------------------------------------------------
-      call get_time (period_time_type, seconds, days)
-      period_out = int(seconds_per_day*days + seconds)
+    call get_time (period_time_type, seconds, days)
+    period_out = int(real(seconds_per_day, r8_kind)*days + seconds)
```
Comment (mlee03) [inline astronomy/astronomy.F90:576]: real(days,r8_kind), real(seconds,r8_kind)

File: astronomy/astronomy.F90 [hunk 19]
```diff
@@ -416,92 +441,109 @@ subroutine astronomy_init (latb, lonb)
...
 !---------------------------------------------------------------------
 !>    Set up time-type variable defining length of year.
 !----------------------------------------------------------------------
-      if (period == 0) then
+    if (period == 0) then
         period_time_type = length_of_year()
         call get_time (period_time_type, seconds, days)
-        period = int(seconds_per_day*days + seconds)
-      else
+            period = int(real(seconds_per_day, r8_kind)*days + seconds)
```
Comment (mlee03) [inline astronomy/astronomy.F90:493]: real(days,r8_kind), real(seconds,r8_kind)

File: astronomy/astronomy.F90 [hunk 20]
```diff
@@ -416,92 +441,109 @@ subroutine astronomy_init (latb, lonb)
...
-      if (present(latb)) then
-        jd = size(latb,2) - 1
-        id = size(lonb,1) - 1
-        allocate (cosz_ann(id, jd))
-        allocate (solar_ann(id, jd))
-        allocate (fracday_ann(id, jd))
-        total_pts = jd*id
-      endif
+    ! check that no invalid types (integers or characters) are given as optional arg
+
```
Comment (mlee03) [inline astronomy/astronomy.F90:518]: is_valid=.false.

File: astronomy/astronomy.F90 [hunk 21]
```diff
@@ -416,92 +441,109 @@ subroutine astronomy_init (latb, lonb)
...
+        type is (real(r4_kind))
+          is_valid = .true.
+        end select
+      type is (real(r8_kind))
+      select type (lonb)
+        type is (real(r8_kind))
+        is_valid = .true.
+      end select
+    end select
+          jd = size(latb,2) - 1
```
Comment (mlee03) [inline astronomy/astronomy.F90:532]: if(is_valid) then 
 id = size(lonb,1) - 1
            allocate (cosz_ann(id, jd))
            allocate (solar_ann(id, jd))
            allocate (fracday_ann(id, jd))
            total_pts = jd*id
end if

**Review by mlee03 — CHANGES_REQUESTED** [review]

File: test_fms/astronomy/test_astronomy.F90 [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+
+program test_astronomy
+
+  use astronomy_mod
+  use fms_mod,           only: fms_init, fms_end
+  use mpp_mod,           only: mpp_error, FATAL
+  use time_manager_mod,  only: JULIAN, set_calendar_type, time_type, set_time
+  use constants_mod,     only: PI
+  use platform_mod,      only: r4_kind, r8_kind
+
```
Comment (mlee03) [inline test_fms/astronomy/test_astronomy.F90:36]: can you add more comments?

File: astronomy/astronomy.F90 (hunk 19 above)
Comment (mlee03) [reply astronomy/astronomy.F90:493]: missed this one?

File: astronomy/astronomy.F90 [hunk 23]
```diff
@@ -416,92 +441,110 @@ subroutine astronomy_init (latb, lonb)
...
         period_time_type = set_time(period,0)
-      endif
+    endif
 
 !---------------------------------------------------------------------
 !>    Define useful module variables.
 !----------------------------------------------------------------------
-      twopi = 2.*PI
-      deg_to_rad = twopi/360.
+    twopi = 2.0_r8_kind * real(PI, r8_kind)
```
Comment (mlee03) [inline astronomy/astronomy.F90:501]: don't need the real() for PI here

File: astronomy/astronomy.F90 [hunk 24]
```diff
@@ -416,92 +441,110 @@ subroutine astronomy_init (latb, lonb)
...
+        type is (real(r4_kind))
+          is_valid = .true.
+        end select
+      type is (real(r8_kind))
+      select type (lonb)
+        type is (real(r8_kind))
+        is_valid = .true.
+      end select
+    end select
+          jd = size(latb,2) - 1
```
Comment (mlee03) [inline astronomy/astronomy.F90:533]: you need a if(is_valid) statement here

**Review by mlee03 — APPROVED** [review]

**Review by rem1776 — APPROVED** [review]

File: astronomy/astronomy.F90 [hunk 25]
```diff
@@ -774,1476 +730,132 @@ subroutine get_ref_date_of_ae (day_out,month_out,year_out,&
...
+type(time_type), intent(in) :: time !< time (1 year = 2*pi) since autumnal equinox
+real(kind=r8_kind)    :: t
 
-!-------------------------------------------------------------------
-!   local variables
-!-------------------------------------------------------------------
-      real, dimension(size(lat),1) :: lat_2d, lon_2d, cosz_2d, &
-                                      fracday_2d, halfday_2d
+    t = (time - autumnal_eq_ref)//period_time_type
+    t = twopi*(t - real(t, r8_kind))
```
Comment (mlee03) [inline astronomy/astronomy.F90:834]: @mcallic2 , was this originally a floor?

Comment (mlee03) [reply astronomy/astronomy.F90:834]: I think it should have been real(floor(t),r8_kind)

**Pull request merged**